### ISSUE TITLE
Provide pnpm workflow caching without action enhancement

### ISSUE THREAD

**Issue by MikeMcC399** [cypress-io/github-action#1227]
- Relates to discussions in https://github.com/cypress-io/github-action/issues/1044

## What would you like

Provide tested recommendations and examples for [pnpm](https://pnpm.io/) workflow caching in the absence of built-in pnpm workflow caching in  `cypress-io/github-action`.

This should be implemented in the workflows:

- [.github/workflows/example-basic-pnpm.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-basic-pnpm.yml)
- [.github/workflows/example-start-and-pnpm-workspaces.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-start-and-pnpm-workspaces.yml)

and documented in the [README > pnpm](https://github.com/cypress-io/github-action/blob/master/README.md#pnpm) section.

## Why is this needed

- Lack of pnpm caching can cause a performance impact on workflows, negatively affecting the usability of `cypress-io/github-action`.
- PR https://github.com/cypress-io/github-action/pull/1188 removed pnpm caching due to reliability issues.
- Although the title of the original community-contribution PR https://github.com/cypress-io/github-action/pull/586 suggested that pnpm dependencies were being cached, in fact there was no additional caching of the [pnpm](https://pnpm.io/) store implemented in this PR.
- The enhancement request https://github.com/cypress-io/github-action/issues/1044 remains open.

**Comment by MikeMcC399** [comment]
The [pnpm documentation](https://pnpm.io/continuous-integration#github-actions) recommends the following:

## GitHub Actions

On GitHub Actions, you can use pnpm for installing and caching your dependencies
like so (belongs in `.github/workflows/NAME.yml`):

```yaml title=".github/workflows/NAME.yml"
name: pnpm Example Workflow
on:
  push:

jobs:
  build:
    runs-on: ubuntu-22.04
    strategy:
      matrix:
        node-version: [20]
    steps:
    - uses: actions/checkout@v4
    - name: Install pnpm
      uses: pnpm/action-setup@v4
      with:
        version: 9
    - name: Use Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'pnpm'
    - name: Install dependencies
      run: pnpm install
```

**Comment by MikeMcC399** [comment]
- The previous recommendation for pnpm caching came from an earlier recommendation of the pnpm documentation.

- Testing again with the relevant commit 8c3db3a4795a005ea7de1f04087c97a3b9ddda84 before caching was removed through PR #1188 shows that it sometimes works and sometimes does not work. When the `example-basic-pnpm` fails, the Linux workflows are affected, and there are 3 of them running in parallel. The failure mechanism is however not clear. Before the test all GitHub Action caches were cleared.

https://github.com/cypress-io/github-action/blob/8c3db3a4795a005ea7de1f04087c97a3b9ddda84/README.md#pnpm

```yaml
name: example-basic-pnpm
on: push
jobs:
  basic-pnpm:
    runs-on: ubuntu-22.04
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Install pnpm
        run: npm install -g pnpm@9
      - name: Get pnpm store directory
        shell: bash
        run: |
          echo "STORE_PATH=$(pnpm store path --silent)" >> $GITHUB_ENV
      - name: Setup pnpm cache
        uses: actions/cache@v4
        with:
          path: ${{ env.STORE_PATH }}
          key: ${{ runner.os }}-pnpm-store-${{ hashFiles('examples/basic-pnpm/pnpm-lock.yaml') }}
      - name: Cypress run
        uses: cypress-io/github-action@v6
        with:
          working-directory: examples/basic-pnpm
```

**Comment by MikeMcC399** [comment]
- PR #1228 resolves this issue as far as caching for pnpm in non-workspace setups is concerned.
- Currently `cypress-io/github-action` does not inherently support pnpm workspaces. See enhancement request https://github.com/cypress-io/github-action/issues/1144. This leads to issues when attempting to apply the pnpm caching recommendations. I will follow this up with separate issue descriptions.

**Comment by MikeMcC399** [comment]
- I successfully verified a test version of the workflow [.github/workflows/example-start-and-pnpm-workspaces.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-start-and-pnpm-workspaces.yml) modified to run mixed older Cypress versions and using `actions/setup-node` for pnpm caching. For the moment I will not update the production version of the workflow [.github/workflows/example-start-and-pnpm-workspaces.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-start-and-pnpm-workspaces.yml) to use caching. It is still a valid example without caching.

- On first run of [.github/workflows/example-start-and-pnpm-workspaces.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-start-and-pnpm-workspaces.yml) `cypress-io/github-action` attempts to verify the installation of Cypress using `npx cypress verify`. Since there is no Cypress installed in the root of the workspace it downloads the latest version of Cypress, which is then cached in the npm cache for subsequent runs. This issue is similar to https://github.com/cypress-io/github-action/issues/971.

- PR https://github.com/cypress-io/github-action/pull/1228 already added pnpm caching to [.github/workflows/example-basic-pnpm.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-basic-pnpm.yml) and updated the [README > pnpm](https://github.com/cypress-io/github-action/blob/master/README.md#pnpm) section to describe pnpm caching. So, I'm now closing this issue as resolved.
